base/linalg/triangular.jl

```diff
@@ -1867,48 +1867,53 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
 end
 logm(A::LowerTriangular) = logm(A.').'
 
-function sqrtm{T}(A::UpperTriangular{T})
-    n = checksquare(A)
+function sqrtm(A::UpperTriangular)
     realmatrix = false
     if isreal(A)
         realmatrix = true
-        for i = 1:n
+        for i = 1:checksquare(A)
             if real(A[i,i]) < 0
                 realmatrix = false
                 break
             end
         end
     end
-    if realmatrix
-        TT = typeof(sqrt(zero(T)))
-    else
-        TT = typeof(sqrt(complex(-one(T))))
-    end
-    R = zeros(TT, n, n)
-    for j = 1:n
-        R[j,j] = realmatrix?sqrt(A[j,j]):sqrt(complex(A[j,j]))
-        for i = j-1:-1:1
-            r = A[i,j]
-            for k = i+1:j-1
-                r -= R[i,k]*R[k,j]
+    sqrtm(A,Val{realmatrix})
+end
+function sqrtm{T,realmatrix}(A::UpperTriangular{T},::Type{Val{realmatrix}})
+    n = checksquare(A)
+    t = realmatrix ? typeof(sqrt(zero(T))) : typeof(sqrt(complex(zero(T))))
+    R = zeros(t, n, n)
+    tt = typeof(zero(t)*zero(t))
+    @inbounds begin
+        for j = 1:n
+            R[j,j] = realmatrix ? sqrt(A[j,j]) : sqrt(complex(A[j,j]))
+            for i = j-1:-1:1
+                r::tt = A[i,j]
+                @simd for k = i+1:j-1
+                    r -= R[i,k]*R[k,j]
+                end
+                r==0 || (R[i,j] = r / (R[i,i] + R[j,j]))
             end
-            r==0 || (R[i,j] = r / (R[i,i] + R[j,j]))
         end
     end
     return UpperTriangular(R)
 end
 function sqrtm{T}(A::UnitUpperTriangular{T})
     n = checksquare(A)
-    TT = typeof(sqrt(zero(T)))
-    R = zeros(TT, n, n)
-    for j = 1:n
-        R[j,j] = one(T)
-        for i = j-1:-1:1
-            r = A[i,j]
-            for k = i+1:j-1
-                r -= R[i,k]*R[k,j]
+    t = typeof(sqrt(zero(T)))
+    R = eye(TT, n, n)
+    tt = typeof(zero(t)*zero(t))
+    one = R[1,1]
+    @inbounds begin
+        for j = 1:n
+            for i = j-1:-1:1
+                r::tt = A[i,j]
+                @simd for k = i+1:j-1
+                    r -= R[i,k]*R[k,j]
+                end
+                r==0 || (R[i,j] = r/one)
```

Comment:
Ah right. Missed the `UnitUpperTriangular` since it was on the other side of a long conversation.